x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationMetrics.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationResult;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+
+/**
+ * Provides a common way of registering and collecting various authentication metrics.
+ *
+ * @param <C> The type of context object which is used to attach additional attributes to collected metrics.
+ */
+final class AuthenticationMetrics<C> {
```

Comment:
I'm open to better name suggestions.

Reply:
Is it possible to split the concerns the of this class such that there is one class for state and another for the wrapping ? Additionally, make both the state and wrapping classes abit more accessible for alternative metrics, such as Authc, Authz, RoleResolution, etc. where the outcome is that if we want to add new concepts is new methods not new classes. 

High level thoughts...
Keep AuthenticationMetrics<C> but rename to SecurityInstrumentation<C> (or SecurityMetrics) create an enumeration for each concept we want to intstrument, i.e. enum {Authc, Authz, Role,  etc.}. The Constructor would accept that enum (in addition to existing constructor args) and would internally map that concept to strings in use here allowing the same code to execute for authz by simply adding new enum with corresponding strings. Change the inner method names to just recordSuccess/Failure. 

Split out the `wrap` method to it's own static class (InstrumentedSecurityActionListener) and change the method name to `wrapForAuthc` , which leaves room to add `wrapForAuthz` or what ever concept we want to instrument. Pass in the SecurityInstrumentation object to the `wrapForAuthc` method and assert that the SecurityInstrumentation enum type is for Authc. 

I think splitting the state / static concerns is a bit cleaner and adopting something a bit more generic now will help set a better foundation and can be more exstisible.

Reply:
> I think splitting the state / static concerns is a bit cleaner and adopting something a bit more generic now will help set a better foundation and can be more exstisible.

Thanks for the suggestions. I liked the proposal and pushed new changes https://github.com/elastic/elasticsearch/pull/103178/commits/7e4d6c30f9be7dd2417d4126e811344ca59bbef1 which attempts to make a better separation of concerns. This should help us to easily add new authentication and authorization metrics. But I'm not so sure (at this point) if it will cover all cases where we plan to collect metrics. Anyway, it's for sure a good start and we can always refactor as we go. Let me know what you think about the refactoring.